addons/mrp/models/stock_rule.py

```diff
@@ -73,7 +73,24 @@ def _run_manufacture(self, procurements):
                     domain += (('procurement_group_id', '=', group.id),)
                 mo = self.env['mrp.production'].sudo().search(domain, limit=1)
             if not mo:
-                new_productions_values_by_company[procurement.company_id.id]['values'].append(rule._prepare_mo_vals(*procurement, bom))
+                if procurement.values.get('batch_size') and procurement.values.get('batch_size') > 0:
+                    total_qty = procurement.product_qty
+                    while total_qty > 0:
+                        qty = min(total_qty, procurement.values['batch_size'])
+                        total_qty -= qty
+                        new_productions_values_by_company[procurement.company_id.id]['values'].append(rule._prepare_mo_vals(
+                            procurement.product_id,
+                            qty,
+                            procurement.product_uom,
+                            procurement.location_id,
+                            procurement.name,
+                            procurement.origin,
+                            procurement.company_id,
+                            procurement.values,
+                            bom
+                        ))
```

Comment:
I suggest something like:
```python
if not mo:
    procurement_qty = procurement.product_qty
    batch_size = procurement.values.get('batch_size', procurement_qty)
    vals = rule._prepare_mo_vals(*procurement, bom)
    while float_compare(procurement_qty, 0, precision_rounding=procurement.product_uom.rounding) > 0:
        current_qty = min(procurement_qty, batch_size)
        new_productions_values_by_company[procurement.company_id.id]['values'].append({
            **vals,
            'product_qty': current_qty
        })
        new_productions_values_by_company[procurement.company_id.id]['procurements'].append(procurement)
        procurement_qty -= current_qty
```
Use float_compare is order to compare float in odoo. To avoid rounding issue with UoM specific rounding.
Always have new_productions_values_by_company['values'] and new_productions_values_by_company['procurements'] should have the same size. However it could be problematic later with 
```
for production, procurement in zip(self, procurements):
```

Reply:
Also we don't need to call _prepare_mo_vals in each loop since the values will be the same except quantity